logical: fix spurious unique constraint errors

jeffswenson opened the pull request:
### logical: refresh value if not provided by cput failure


This reworks the LDR KV writer so that it explicitly issues a read
if the cput error belongs to a non-primary key index. This fixes an
issue where replays could cause spurious DLQ entries.

Release note: Fixes a bug in LDR that resulted in DLQing rows if a
unique secondary index has an index id less than the primary key index.
Informs: https://github.com/cockroachdb/cockroach/issues/145994

### logical: remove partial batch support from kv writer

This simplifies the kv writer by removing some of the transaction
handling that was written to support batches.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146465)
<!-- Reviewable:end -->

jeffswenson committed:
logical: remove partial batch support from kv writer

This simplifies the kv writer by removing some of the transaction
handling that was written to support batches.

Release note: none

jeffswenson committed:
logical: refresh value if not provided by cput failure

This reworks the LDR KV writer so that it explicitly issues a read
if the cput error belongs to a non-primary key index. This fixes an
issue where replays could cause spurious DLQ entries.

Release note: Fixes a bug in LDR that resulted in DLQing rows if a
unique secondary index has an index id less than the primary key index.
Informs: #145994

jeffswenson commented:
The race failure seems to suggest that using a `scratch` buffer for a `Get` key is not allowed. Even if the scratch buffer isn't reused across transactions 🤔

jeffswenson commented:
Closing in favor of #147117
